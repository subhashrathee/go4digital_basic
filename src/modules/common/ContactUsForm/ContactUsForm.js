//eslint-disable jsx-a11y/alt-text
import React from 'react';

//Config object used for emailjs vars
const Config = require('../../config/config.json');

//Styles used by html input elements
const styles = {
          error:  {
            "color"    : "red",
            "fontSize" : "12px",
            "marginTop": "5px"
          },
          hidden: {
            "display"  : "none"
          } 
      };

//The email validation regular expression we check against later.
//We ignore this error because it's a regex, not ES error (but linter picking it up).
// eslint-disable-next-line
const emailRegEx = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

/**
* Contact us form component used to render and process the contact us form.
*
* Props: None
*
* State vars:
* @String email   The email address entered by the user.
* @String message The message entered by the user.
* @String emailErrorMessage The email error. This is updated if there's an error or not.
* @String messageErrorMessage The message error. This is updated if there's an error or not.1
*/
class ContactUsForm extends React.Component {

  /**
  * Default constructor. Sets up the state and methods we need bound.
  */
  constructor(props) {
    super(props);
    
    this.state = { email              : '',
                   message            : '',
                   forbiddenInput     : '',
                   emailErrorMessage  : '',
                   messageErrorMessage: '' };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit      = this.handleSubmit.bind(this);
  }

  /**
  * Method to handle the onChange events.
  * It calls the showInputError to validate the error input as the user 
  * enters a value.
  *
  * @param Object event The event to get the target element.
  * 
  * @return void
  */
  handleInputChange(event) { 

    const target = event.target;

    target.classList.add('active');

    this.setState({
      [target.name]: target.value
    });

    this.showInputError(target.name);
  }


  /**
  * Method to handle the subimt events.
  * It checks if the form is valid by calling the showFormErrors method.
  * It then forces an update so either we show the error messages or the
  * thankyou message.
  * 
  * @param Object event The event to get the target element.
  * 
  * @return void
  */
  handleSubmit(event) {
    event.preventDefault();
 
    var isValid = this.showFormErrors();
 
    if(isValid) {

      if( this.isNotSpamSubmission() ) {

        document.getElementById('thankYouMessage').style="display:inline;"
        document.getElementById('form').style="display:none;"

        window.emailjs.init(Config.emailjs_user);
        window.emailjs.send(Config.emailjs_service_id, Config.emailjs_template_id,
                    {email  : this.refs.email.value, 
                     message: this.refs.message.value});
      }
    }

    this.forceUpdate();

  }	

  /**
  * isSpamSubmission checks for a hidden field which should not be set and returns false if it is
  *
  * @param string spamField
  *
  * @return boolean Returns true if no spam
  */
  isNotSpamSubmission() {
  
    var spamFieldValue = this.state.forbiddenInput;

    if(spamFieldValue.length !== 0) {
      return false;
    }
    
    return true; 
  }

  /**
  * Shows the form errors (triggered by an onclick of the form).
  * It gets all the inputs and text areas (add more if you need), and 
  * uses the showInputError method to validate each field. 
  * 
  * @return Boolean Returns true if the form is valid, otherwise false if an error is found.
  */
  showFormErrors() {
    const inputs = document.querySelectorAll('input[type=email], textarea');

    let isFormValid = true;
    
    inputs.forEach(input => {
      input.classList.add('active');
      
      if (!this.showInputError(input.name)) {
        isFormValid = false;
      }
    });

    return isFormValid;
  }

  /**
  * showInputError method which shows the errors associated with the 
  * ref passed in. Called by onChange event for each input and showFormErrors
  * which is called by handleSubmit.
  * 
  * @param ref string The reference of the input which called it.
  */
  showInputError(ref) {
    var validity = this.refs[ref].validity;
    var label    = document.getElementById(`${ref}Label`).textContent;
    
    var newState = this.state;
    var errorString = '';      
    //This is a silly error. We are using it, just inside setState
    // eslint-disable-next-line
    var errorMessageField = `${ref}ErrorMessage`;

    const isEmail   = ref.indexOf('email') !== -1;

    //set custom validity for email, doesn't work on all browsers
    if(isEmail) {
      if(!emailRegEx.test(this.refs.email.value.toLowerCase())) {
        validity.email = false;
      } else {
        validity.email = true;
      }
    }
 
    //if our form isn't valid, then find what the error is and return the correct
    //message to the user. 
    if (!validity.valid) {

      if (validity.valueMissing) {
        errorString = `${label} is a required field`
      } else if (isEmail && !validity.email) {
        errorString = `${label} is not a valid email address`; 
      } 

      //dynamically update key in state
      newState[errorMessageField] = errorString;
      this.setState({newState});

      return false;
    }

    //if no errors, blank out the error messages
    newState[errorMessageField] = '';
    this.setState({newState});
    return true;
  }

  /**
  * Render method. This is used to render the form and the thankyou message.
  * Onchange events on the inputs validate the inputs via handleInputChange method and
  * there's a handleSubmit event.
  *
  * @return void
  */
  render() {
    return (
      <div className="main">
        <div id="thankYouMessage" style={styles.hidden}>
          <p>Thanks for submitting your enquiry. We'll get back to you shortly</p>
        </div> 
          <form id="form">
        <p>Call Sarita Singh for any enquiries on: 0414 043 200</p>
            <label id="emailLabel" style={styles.hidden}>
              Email
            </label>
            <input type="email" placeholder="Email address" value={this.state.email} name="email" id="email" ref="email" onChange={this.handleInputChange} required/>
            <div className="error" id="emailError" style={styles.error}>{this.state.emailErrorMessage}</div>
            <br />
            <label id="messageLabel" style={styles.hidden}>
              Message
            </label>
            <textarea rows="5" cols="30" placeholder="Message" value={this.state.message} name="message" id="message" ref="message" onChange={this.handleInputChange} required></textarea>
            <div className="error" id="messageError" style={styles.error}>{this.state.messageErrorMessage}</div>
            <br />
            <label id="forbiddenInputLabel" style={styles.hidden}></label>
            <input type="input" style={styles.hidden} name="forbiddenInput" id="forbiddenInput" ref="forbiddenInput" value={this.state.forbiddenInput} onChange={this.handleInputChange} />
            <input type="submit" name="submit" value="Submit" onClick={this.handleSubmit} />
        </form>
      </div>
    )
  }

}


export default ContactUsForm;
