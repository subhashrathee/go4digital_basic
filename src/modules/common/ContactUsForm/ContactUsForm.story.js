import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { BrowserRouter } from 'react-router-dom'
import ContactUsForm from './ContactUsForm'

storiesOf('ContactUsForm', module)
  .add('Default Contact Us Form', () => (
      <ContactUsForm />
  ));
