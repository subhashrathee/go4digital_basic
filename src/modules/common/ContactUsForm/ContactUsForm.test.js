import React from 'react';
import ContactUsForm from './ContactUsForm';
import renderer from 'react-test-renderer';
import { BrowserRouter, Link} from 'react-router-dom'
import { shallow, configure, mount, find, ref, findWhere } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('<MyComponent />', () => {

  it('renders a contctUsFormComponent', () => {
    const component = renderer.create(
      <ContactUsForm />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  //handleInputChange event
  it('handles on input change event', () => {

    const output = mount(<ContactUsForm />);

    //when showInputError(ref) is called, simulate return false
    //use a stub
    output.instance().showInputError = jest.fn();
    output.update();

    //manually mock the event we pass in
    var mockEvent = {};
    mockEvent.target = new Object();
    mockEvent.target.classList = new Object();
    mockEvent.target.name = {validity:false};
    mockEvent.target.value = "John@test.com";

    //mock the target object
    mockEvent.target.classList.add = jest.fn();
   
    //our assertions
    expect(output.state().email).toEqual('');
  
    //find email input and test it for value change (with validity = false)
    output.state().email = mockEvent.target.value;
    output.find('input').first().simulate('change', {target:  mockEvent.target });
    expect(output.state().email).toEqual('John@test.com');

    //console.log(output.debug());
    //console.log(output.find('input').debug());

    mockEvent.target.name = {validity:true}; 
    //find email input and test it for value change (with validity = true)
    output.state().email = mockEvent.target.value;
    output.find('input').first().simulate('change', {target:  mockEvent.target });
    expect(output.state().email).toEqual('John@test.com');

    //handleinputchange for forbidden field
    output.state().forbiddenField = mockEvent.target.value;
  });

  //showInputError method
  it('handles showInputError method', () => {

    const output = mount(<ContactUsForm />, { attachTo: document.body });

    //mock the email validity ref
    output.ref('email').validity = new Object;

    expect(output.instance().showInputError('email')).toEqual(false);

    //mock the email ref value
    output.ref('email').value = 'valid@test.com';

    expect(output.instance().showInputError('email')).toEqual(false);

    //mock the validity valueMissing value
    output.ref('email').validity.valueMissing = true;

    expect(output.instance().showInputError('email')).toEqual(false);
  });

  //showFormErrors method
  it('handles showFormErrors', () => {
    const output = mount(<ContactUsForm />); 

    output.ref('email').validity = new Object;
    output.ref('email').validity.valid = false;
    output.ref('message').validity = new Object;
    output.ref('message').validity.valid = false;

    output.instance.showInputError = jest.fn();
    output.update();

    expect(output.instance().showFormErrors()).toEqual(false);
  });

  it('handles handlesubmit', () => {
    
    const output = mount(<ContactUsForm />);

    output.instance.showFormErrors = false;

    const eventVar = new Object();
    eventVar.preventDefault = jest.fn();

    //mock the input ref validity values
    output.ref('email').validity = new Object;
    output.ref('email').validity.valid = false;
    output.ref('message').validity = new Object;
    output.ref('message').validity.valid = false;

    //isValid is set to false, so expecting false
    output.instance().handleSubmit(eventVar)
   
    //compare the 2nd div (the thankyoumessage one)
    expect(output.find('div').at(1).prop('style').display).toEqual('none');

    //now we've tested with false, lets test a positive case

    //remock the email and message validity objects
    output.ref('email').validity.valid = true;
    output.ref('message').validity.valid = true;

    //mock the window emailjs functions to send emails.
    global.emailjs = new Object;
    global.emailjs.init = jest.fn();
    global.emailjs.send = jest.fn();

    //isValid is set to false, so expecting false
    output.instance().handleSubmit(eventVar)
    output.update();

    //compare the 2nd div (the thankyoumessage one)
    expect(output.find('div').at(1).prop('style').display).toEqual('none');
    //compare the next div to ensure the form is hidden

    //testing for hidden spam field, send emails if it's not populated
    expect(output.state().forbiddenInput).toEqual('');

    //don't send emails if it's populated
    output.state().forbiddenInput = 'test';
    expect(output.state().forbiddenInput).toEqual('test');
  });
  
  /**
  * Test the isNotSpamSubmission method
  */
  it('isNotSpamSubmission should work', () => {
    const output = mount(<ContactUsForm />);

    //test method with no data in it.
    output.state().forbiddenInput = '';
    expect(output.instance().isNotSpamSubmission()).toEqual(true);

    //test method with data in it.
    output.state().forbiddenInput = 'test';
    expect(output.instance().isNotSpamSubmission()).toEqual(false);

  });
});
