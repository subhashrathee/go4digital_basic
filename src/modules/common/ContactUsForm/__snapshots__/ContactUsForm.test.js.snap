// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyComponent /> renders a contctUsFormComponent 1`] = `
<div
  className="main"
>
  <div
    id="thankYouMessage"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <p>
      Thanks for submitting your enquiry. We'll get back to you shortly
    </p>
  </div>
  <form
    id="form"
  >
    <p>
      Call Sarita Singh for any enquiries on: 0414 043 200
    </p>
    <label
      id="emailLabel"
      style={
        Object {
          "display": "none",
        }
      }
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Email address"
      required={true}
      type="email"
      value=""
    />
    <div
      className="error"
      id="emailError"
      style={
        Object {
          "color": "red",
          "fontSize": "12px",
          "marginTop": "5px",
        }
      }
    >
      
    </div>
    <br />
    <label
      id="messageLabel"
      style={
        Object {
          "display": "none",
        }
      }
    >
      Message
    </label>
    <textarea
      cols="30"
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="Message"
      required={true}
      rows="5"
      value=""
    />
    <div
      className="error"
      id="messageError"
      style={
        Object {
          "color": "red",
          "fontSize": "12px",
          "marginTop": "5px",
        }
      }
    >
      
    </div>
    <br />
    <label
      id="forbiddenInputLabel"
      style={
        Object {
          "display": "none",
        }
      }
    />
    <input
      id="forbiddenInput"
      name="forbiddenInput"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="input"
      value=""
    />
    <input
      name="submit"
      onClick={[Function]}
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
